pkg/trace/agent/agent_test.go

```diff
@@ -581,3 +586,165 @@ func formatTrace(t pb.Trace) pb.Trace {
 	}
 	return t
 }
+
+func BenchmarkThroughput(b *testing.B) {
+	if _, ok := os.LookupEnv("DD_TRACE_TEST_PRIVATE_DATA"); !ok {
+		b.SkipNow()
+	}
+
+	ddlog.SetupDatadogLogger(seelog.Disabled, "") // disable logging
+
+	b.Run("39KB/10", benchThroughput("src/github.com/DataDog/trace-agent-test/payloads/10traces.msgp"))
+	b.Run("1.1MB/100", benchThroughput("src/github.com/DataDog/trace-agent-test/payloads/100traces.msgp"))
+	b.Run("1.1MB/1951", benchThroughput("src/github.com/DataDog/trace-agent-test/payloads/26.msgp"))
+	b.Run("2.2MB/4161", benchThroughput("src/github.com/DataDog/trace-agent-test/payloads/10.msgp"))
+	b.Run("6.6MB/500", benchThroughput("src/github.com/DataDog/trace-agent-test/payloads/500traces.msgp"))
+}
+
+func benchThroughput(file string) func(*testing.B) {
+	return func(b *testing.B) {
+		data, count, err := tracesFromFile(file)
+		if err != nil {
+			b.Fatal(err)
+		}
+		cfg := config.New()
+		cfg.Endpoints[0].APIKey = "irrelevant"
+
+		ctx, cancelFunc := context.WithCancel(context.Background())
+		http.DefaultServeMux = &http.ServeMux{}
+		agnt := NewAgent(ctx, cfg)
+		defer cancelFunc()
+
+		// start the agent without the trace writer; we will be draining
+		// the write channel ourselves in the benchmarks.
+		startAgentWithoutTraceWriter(ctx, b, agnt)
+
+		// drain every other channel to avoid blockage.
+		exit := make(chan bool)
+		go func() {
+			defer close(exit)
+			for {
+				select {
+				case <-agnt.ServiceExtractor.outServices:
+				case <-agnt.Concentrator.OutStats:
+				case <-exit:
+					return
+				}
+			}
+		}()
+
+		b.ResetTimer()
+		b.SetBytes(int64(len(data)))
+
+		for i := 0; i < b.N; i++ {
+			req, err := http.NewRequest("PUT", "http://localhost:8126/v0.4/traces", bytes.NewReader(data))
```

Comment:
Nit: How much of the request creation could we extract out of the benchmark loop? Not that important but that'd help strip time/allocs from the bench which wouldn't be representative of the Agent performance.

Reply:
I could replace `req.Body` and keep the same instance, but I'm skeptical about if that'd be correct. I don't think this uses that much to make it significant. Out of the total this is such a small percentage that it wouldn't really matter imho.